python/lsst/ts/standardscripts/base_close_loop.py

```diff
@@ -0,0 +1,499 @@
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+__all__ = ["BaseCloseLoop"]
+
+import abc
+import types
+import typing
+
+import numpy as np
+import yaml
+from lsst.ts import salobj
+from lsst.ts.observatory.control.maintel.mtcs import MTCS, MTCSUsages
+from lsst.ts.observatory.control.utils.enums import ClosedLoopMode, DOFComponent
+
+STD_TIMEOUT = 10
+
+
+class BaseCloseLoop(salobj.BaseScript, metaclass=abc.ABCMeta):
+    """Closed loop script. This script is used to perform measurements of
+    the wavefront error, then propose dof offsets based on ts_ofc.
+
+    Parameters
+    ----------
+    index : `int`, optional
+        Index of Script SAL component (default=1).
+    remotes : `bool`, optional
+        Should the remotes be created (default=True)? For unit testing this
+        can be set to False, which allows one to mock the remotes behaviour.
+    descr : `str`, optional
+        Short description of the script.
+
+    Notes
+    -----
+    **Checkpoints**
+
+    - "Taking image...": If taking in-focus detection image.
+    - "[N/MAX_ITER]: Closed loop starting...": Before each closed loop
+        iteration, where "N" is the iteration number and "MAX_ITER"
+        is the maximum number of iterations.
+    - "[N/MAX_ITER]: Closed converged.": Once Closed Loop reaches convergence.
+    - "[N/MAX_ITER]: Closed applying correction.": Just before
+        corrections are applied.
+
+    **Details**
+
+    This script is used to perform measurements of the wavefront error, then
+    propose dof offsets based on ts_ofc. The offsets are not applied
+    automatically and must be turned on by the user through
+    apply_corrections attribute.
+    """
+
+    def __init__(self, index=1, add_remotes=True, descr=""):
+        super().__init__(
+            index=index,
+            descr=descr,
+        )
+        self.add_remotes = add_remotes
+        self.mtcs = None
+        self.camera = None
+
+        # The following attributes are set via the configuration
+        self.filter = None
+
+        # exposure time for the intra/extra images (in seconds)
+        self.exposure_time = None
+
+        # Define operation mode handler function
+        self.operation_model_handlers = {
+            ClosedLoopMode.CWFS: self.handle_cwfs_mode,
+            ClosedLoopMode.FAM: self.handle_fam_mode,
+        }
+
+    @property
+    @abc.abstractmethod
+    def camera(self):
```

Comment:
Can you add type annotation to these method? I notice below that you did `ComCam | None` and `LSSTCam | None`. Type-checking wise this will make our life harder. We currently don't have type checking on in this repo but looking forward I would like to add support for it. 

I'd recommend you only return the camera type and raise an exception if the camera is `None`.  Something like this:

```
@property
@abc.abstractmethod
def camera(self) -> BaseCamera:
    raise NotImplementedError()
```

Then for ComCam and LSSTCam you would do something like this:

```
@property
def camera(self) -> BaseCamera:
    if self._comcam is not None:
        return self._comcam
    else:
        raise RuntimeError("Camera not defined.")
```

You can either use `BaseCamera` or `ComCam` as the return type in this case.